src/3devoUpdate/Form1.cs

```diff
@@ -1,610 +1,541 @@
+        btnFlashBrowse.Enabled = false;
+        cmbPort.Enabled = false;
+      });
+    }
 
-        // Browse for flash file
-        private void btnFlashBrowse_Click(object sender, EventArgs e)
-        {
-            if (openFileDialog1.ShowDialog() == DialogResult.OK)
-            {
-                txtFlashFile.Text = openFileDialog1.FileName;                
-            }                
+    private void UpdateInterface() {
+      // Clear status info text
+      txtStatusInfo.Clear();
+
+      cmbPort.Enabled = true;
+
+      // No devices connected/found
+      if( cmbPort.Items.Count == 0 ) {
+        btnFlashBrowse.Enabled = false;
+        txtStatusInfo.AppendText(NO_MACHINE_CONNECTED);
+      }
+      // 1 or more devices connected
+      else {
+        btnFlashBrowse.Enabled = true;
+
+        if( deviceSelected == false ) {
+          txtStatusInfo.AppendText(PORT_NOT_SELECTED);
+        } else if( flashFile.Length == 0 ) {
+          txtStatusInfo.AppendText(FILE_NOT_SELECTED);
+        } else {
+          txtStatusInfo.AppendText(READY_FOR_UPLOADING);
         }
+      }
+
+      // User warning when the machine is detected in bootloader mode.
+      if (bootloaderDevice == true) {
+        txtStatusInfo.AppendText("\r\nGeneric ST device selected;\n");
+        txtStatusInfo.AppendText("Carefully select the correct binary file for your machine, since we cannot determine which 3devo device is connected!\n");
+        txtStatusInfo.AppendText("Uploading the wrong firmware may cause hardware damage or the machine might stop responding.\n");
+      }
+
+      if( (deviceSelected == true)
+          && (downloadFileSelected == true)
+          && (!downloadIsReady) ) {
+        downloadIsReady = true;
+        btnUpload.Enabled = true;
+      }
+    }
+
+    private void StartUploadProcess() {
+      switch( DeviceInfoList.ElementAt(cmbPort.SelectedIndex).MachineName ) {
+        case USBDeviceList.MachineType.FilamentMaker:
+          avrCmdLine.generate();
+          avrdude.Launch(avrCmdLine.cmdBox, Avrdude.CommandType.FILE_UPLOAD);
+          break;
+        case USBDeviceList.MachineType.StBootloader: // Fall through
+        case USBDeviceList.MachineType.AiridDryer:
+          USBDeviceList usbDevice;
+          try {
+            // This assumes that the selected combobox index doesn't change (at least not before the upload process).
+            usbDevice = DeviceInfoList.ElementAt(cmbPort.SelectedIndex);
+          }
+          catch( ArgumentOutOfRangeException ) {
+            throw new ArgumentOutOfRangeException("StartUploadProcess(); combobox index doesn't exist: ", cmbPort.SelectedIndex.ToString());
+          }
+
+          // TODO: Place the Target information somewhere else and make it less hardcoded.
+          if( !dfuSeFileValidation.FirmwareFileValidation(usbDevice, "3devo-AiridMainboard", dfuUtilCmdLine.dfuseAddress) ) {
+            EnableInterface(false /*No upload button*/);

```

Comment:
Shouldn't this return an error message to the user? They already clicked upload when they end up here?